airflow-core/tests/unit/jobs/test_triggerer_job.py

```diff
@@ -589,3 +595,125 @@ def test_failed_trigger(session, dag_maker, supervisor_builder):
+        )
+
+
+class DummyTriggerRunnerSupervisor(TriggerRunnerSupervisor):
+    """
+    Make sure that the Supervisor stops after handling the events and do not keep running forever so the
+    test can continue.
+    """
+
+    def handle_events(self):
+        self.stop = bool(self.events)
+        super().handle_events()
+
+
+@pytest.mark.asyncio
+# @patch("airflow.providers.standard.triggers.temporal.CustomTrigger", return_value=CustomTrigger)
+async def test_trigger_can_access_variables_and_connections(session, dag_maker, supervisor_builder):
+    """
+    Checks that the trigger will successfully access Variables, Connections and XCom.
+
+    This is the Supervisor side of the error reported in TestTriggerRunner::test_invalid_trigger
+    """
+
+    # Create the test DAG and task
+    with dag_maker(dag_id="trigger_accessing_variable_and_connection", session=session):
+        EmptyOperator(task_id="dummy1")
+    dr = dag_maker.create_dagrun()
+    task_instance = dr.task_instances[0]
+    # Make a task instance based on that and tie it to the trigger
+    task_instance.state = TaskInstanceState.DEFERRED
+
+    # Create a Trigger
+    trigger = CustomTrigger(dag_id=dr.dag_id, run_id=dr.run_id, task_id=task_instance.task_id, map_index=-1)
+    trigger_orm = Trigger(
+        classpath=trigger.serialize()[0],
+        kwargs={"dag_id": dr.dag_id, "run_id": dr.run_id, "task_id": task_instance.task_id, "map_index": -1},
+    )
+    trigger_orm.id = 1
+    session.add(trigger_orm)
+    session.commit()
+    task_instance.trigger_id = trigger_orm.id
+
+    # Create the appropriate Connection, Variable and XCom
+    connection = Connection(conn_id="test_connection", conn_type="http")
+    variable = Variable(key="test_variable", val="some_variable_value")
+    XComModel.set(
+        key="test_xcom",
+        value="some_xcom_value",
+        task_id=task_instance.task_id,
+        dag_id=dr.dag_id,
+        run_id=dr.run_id,
+        map_index=-1,
+        session=session,
+    )
+    session.add(connection)
+    session.add(variable)
+
+    job = Job()
+    session.add(job)
+    session.commit()
+
+    supervisor = DummyTriggerRunnerSupervisor.start(job=job, capacity=1, logger=None)
+    supervisor.run()
+
+    task_instance.refresh_from_db()
+    assert task_instance.state == TaskInstanceState.SCHEDULED
+    assert task_instance.next_method != "__fail__"
+    assert task_instance.next_kwargs == {
+        "event": {
+            "connection": {
+                "conn_id": "test_connection",
+                "conn_type": "http",
+                "description": None,
+                "host": None,
+                "schema": None,
+                "login": None,
+                "password": None,
+                "port": None,
+                "extra": None,
+            },
+            "variable": "some_variable_value",
+            "xcom": '"some_xcom_value"',
```

Comment:
Just to bring attention there, is that expected. (double quotes `"'`)

Reply:
@amoghrajesh Does this look right to you?

Reply:
It's the result of calling

```
        xcom = await sync_to_async(XCom.get_one)(
            key="test_xcom",
            dag_id=self.dag_id,
            run_id=self.run_id,
            task_id=self.task_id,
            map_index=self.map_index,
        )
```

Reply:
That's due to the mismatch between how we read vs write xcoms here.

We are setting using `XcomModel.set` which serialises into a JSON str and stores it but we are getting the xcom using `Xcom.get` from sdk which doesn't deserialise it (https://github.com/apache/airflow/pull/48239/files#diff-6ad6972d85e00292c6f4b03b5ecd5da89c837c23d6077a1f44cc726a1b95bd5aR633-R639). So we should instead make the call above in `CustomTrigger` to do `XcomModel.get_many` instead which will deserialise it as well as needed.